content/post/idsplit.md

```diff
@@ -0,0 +1,122 @@
+"progressbar" and "control", because the usage contexts are quite different.
+Around 7% of the identifiers in our dataset are not splittable by heuristics,
+and further 15% contain at least one splittable part after applying heuristics.
+That is more than 10 million affected strings.
+
+## Solution
+
+Machine Learning team is going to make machines learn, so we studied several
+ML models to solve the splitting task. Before jumping to the juicy ML stuff,
+we need to define our training dataset, which seems to be absent at first glance.
+However, it is possible to turn our dataset of extracted identifiers into
+(X, Y) by trusting the heuristics. Indeed, we can take the identifiers
+that are splittable by case change, underscore, etc. and pretend that the
+outcome is the ground truth split points. That gives us about 50 million traning
+examples for free.
+
+Of course, as was already said above, some of the heuritical splits are not enough,
+so our ML models are going to be trained on noisy labels. However, we can overcome
+this limitation by bootstrapping the dataset and training several generations
+of the same model - in other words, refining the training examples at each iteration
+by applying the current model to them. We did not research that, unfortunately.
+
+One last remark: we are going to evaluate the models by measuring their
+[precision and recall](https://en.wikipedia.org/wiki/Precision_and_recall).
+Precision equals to how many split points we predicted correctly divided by the
+total number of predictions. Recall equals to how many split points we predicted
+correctly divided by the total number of ground truth splits. They are both
+important and indicate the bias towards spammy, noisy predictions which cover
+everything or conservative, confident ones which trigger on few samples.
+
+Here are the models which we evaluated.
+
+#### Maximum likelihood character-level, unsmoothed
+
+This is a quick and dirty baseline implemented in
+[CharStatModel](https://github.com/vmarkovtsev/CharStatModel) repository. We
+record all possible sequences of characters in form of a tree
+and decide whether to split based on the tree branch with the dominating weight
+for a given prefix or suffix. There are several weak points of that approach:
+it consumes much memory, it is always biased towards good precision or good recall,
+it is quite short-sighted.
+
+#### Dynamic programming
+
+Accepting the hypothesis that the atoms of the string are independent from one
+another, we can model the probability of a sequence of words using frequencies
+computed on a corpus. That corpus can be our training dataset or anything else.
+We explored the former and additionally the Wikipedia corpus. There can be also
+different assumptions about the probability distribution (prior), e.g.
+training posterior or [Ziph](https://en.wikipedia.org/wiki/Zipf%27s_law).
+We tried three models with those combinations.
```

Comment:
Is there anything else to say about these combinations? Why did you go with this one and not others?

Reply:
The curious reader can always read the details in the paper. The explanation is going to increase the volume.